File: Tourism/app/src/main/java/com/huawei/discovertourismapp/fragments/LoginFragment.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2020-2022. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.huawei.discovertourismapp.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.huawei.discovertourismapp.R;
import com.huawei.discovertourismapp.utils.Constants;
import com.huawei.discovertourismapp.utils.Util;
import com.huawei.discovertourismapp.utils.Validation;
import com.huawei.discovertourismapp.viewmodel.AuthServiceViewModel;

public class LoginFragment extends Fragment {
    private LoginFragmentListener loginFragmentListener;
    private AuthServiceViewModel authServiceViewModel;

    public void setLoginFragmentListener(LoginFragmentListener loginFragmentListener) {
        this.loginFragmentListener = loginFragmentListener;
    }
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        authServiceViewModel = new ViewModelProvider(requireActivity()).get(AuthServiceViewModel.class);
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login_phone, container, false);

        Button mOtpBtn = view.findViewById(R.id.otp_btn);
        EditText etPhoneNumber = view.findViewById(R.id.etPhoneNumber);

        mOtpBtn.setOnClickListener(v -> {

            Bundle bundle = new Bundle();
            if (Validation.isValidation(etPhoneNumber.getText().toString(), "+91")) {
                Util util=new Util();
                util.showProgressBar(getActivity());
                bundle.putString(Constants.PHONE_NUMBER, etPhoneNumber.getText().toString());
                bundle.putString(Constants.COUNTRY_CODE, "+91");
                authServiceViewModel.getOTP("+91", etPhoneNumber.getText().toString());
                authServiceViewModel.verifyCodeResultMutableLiveData.observe(requireActivity(), verifyCodeResult -> {
                    loginFragmentListener.setLoginFragmentListener(bundle);
                    util.stopProgressBar();
                });

            } else {
                Toast.makeText(requireContext(), R.string.enter_number, Toast.LENGTH_SHORT).show();
            }

        });
        return view;


    }

    public interface LoginFragmentListener {
        public void setLoginFragmentListener(Bundle bundle);
    }
}

File: HealthyCycling/app/src/main/java/com/myapps/hibike/ui/login/LoginFragment.kt
/*
 * Copyright 2022. Huawei Technologies Co., Ltd. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.myapps.hibike.ui.login

import android.Manifest
import android.app.Activity
import android.app.AlertDialog
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.findNavController
import com.huawei.hms.support.account.service.AccountAuthService
import com.myapps.hibike.R
import com.myapps.hibike.databinding.FragmentLoginBinding
import com.myapps.hibike.utils.extension.goSettings
import com.myapps.hibike.utils.extension.hide
import com.myapps.hibike.utils.extension.show
import com.myapps.hibike.utils.extension.showToast
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class LoginFragment : Fragment() {

    private lateinit var binding: FragmentLoginBinding
    private val loginViewModel: LoginViewModel by viewModels()

    @Inject
    lateinit var authService: AccountAuthService

    private var signInWithHuaweiID =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == Activity.RESULT_OK) {
                loginViewModel.signIn(result.data)
            }
        }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentLoginBinding.inflate(inflater)

        initListeners()
        setupObserver()
        onBackPressed()

        return binding.root
    }

    private fun initListeners() {
        binding.btnSignIn.setOnClickListener {
            if (checkPermission()) signInWithHuaweiID.launch(authService.signInIntent)
            else showPermissionDialog()
        }
    }

    private fun showPermissionDialog(){
        val alertDialog = AlertDialog.Builder(requireActivity())
            .setMessage(getString(R.string.permissions_dialog_text))
            .setNegativeButton(getString(R.string.alert_neg)) { dialog, _ ->
                dialog.dismiss()
            }
            .setPositiveButton(getString(R.string.go_settings)) { _, _ ->
                goSettings()
            }.create()
        alertDialog.show()
    }

    private fun setupObserver() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                loginViewModel.authState.collect { loginUiState ->
                    if (loginUiState.isUserSignedIn && loginUiState.isUserCreated) {
                        binding.loginProgress.hide()
                        findNavController().navigate(R.id.action_loginFragment_to_homeFragment)
                    }
                    if (loginUiState.isLoading) {
                        binding.loginProgress.show()
                    }
                    if (loginUiState.error.isNotEmpty()) {
                        binding.loginProgress.hide()
                        showToast(loginUiState.error)
                    }
                }
            }
        }
    }

    private fun onBackPressed() {
        val callback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                activity?.finish()
            }
        }
        requireActivity().onBackPressedDispatcher.addCallback(viewLifecycleOwner, callback)
    }

    private fun checkPermission(): Boolean {
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
            if (ActivityCompat.checkSelfPermission(
                    requireContext(),
                    Manifest.permission.ACCESS_FINE_LOCATION
                ) != PackageManager.PERMISSION_GRANTED
                || ActivityCompat.checkSelfPermission(
                    requireContext(),
                    Manifest.permission.ACCESS_COARSE_LOCATION
                ) != PackageManager.PERMISSION_GRANTED
                || ActivityCompat.checkSelfPermission(
                    requireContext(),
                    Manifest.permission.CAMERA
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                return false
            }
        } else {
            if (ActivityCompat.checkSelfPermission(
                    requireContext(),
                    Manifest.permission.ACCESS_FINE_LOCATION
                ) != PackageManager.PERMISSION_GRANTED || ActivityCompat.checkSelfPermission(
                    requireContext(),
                    Manifest.permission.ACCESS_COARSE_LOCATION
                ) != PackageManager.PERMISSION_GRANTED || ActivityCompat.checkSelfPermission(
                    requireContext(),
                    Manifest.permission.CAMERA
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                return false
            }
        }
        return true
    }
}